src/Storages/MergeTree/ReplicatedMergeTreeMergeStrategyPicker.cpp

```diff
@@ -0,0 +1,151 @@
+#include <Storages/MergeTree/ReplicatedMergeTreeMergeStrategyPicker.h>
+#include <Storages/StorageReplicatedMergeTree.h>
+#include <Storages/MergeTree/ReplicatedMergeTreeLogEntry.h>
+
+
+#include <common/types.h>
+#include <optional>
+#include <mutex>
+#include <city.h>
+#include <algorithm>
+
+namespace DB
+{
+
+// minimum interval (seconds) between checks if chosen replica finished the merge.
+static const auto RECHECK_MERGE_READYNESS_INTERVAL_SECONDS = 1;
+
+// don't refresh state too often (to limit number of zookeeper ops)
+static const auto REFRESH_STATE_MINIMUM_INTERVAL_SECONDS = 5;
+
+// refresh the state automatically it it was not refreshed for a longer time
+static const auto REFRESH_STATE_MAXIMUM_INTERVAL_SECONDS = 30;
+
+
+ReplicatedMergeTreeMergeStrategyPicker::ReplicatedMergeTreeMergeStrategyPicker(StorageReplicatedMergeTree & storage_)
+    : storage(storage_)
+{}
+
+
+bool ReplicatedMergeTreeMergeStrategyPicker::isMergeFinishedByReplica(const String & replica, const ReplicatedMergeTreeLogEntryData & entry)
+{
+    // those have only seconds resolution, so recheck period is quite rough
+    auto reference_timestamp = entry.last_postpone_time;
+    if (reference_timestamp == 0)
+        reference_timestamp = entry.create_time;
+
+    // we don't want to check zookeeper too frequent
+    if (time(nullptr) - reference_timestamp >= RECHECK_MERGE_READYNESS_INTERVAL_SECONDS)
+    {
+        return storage.checkReplicaHavePart(replica, entry.new_part_name);
+    }
+
+    return false;
+}
+
+
+// that will return the same replica name for ReplicatedMergeTreeLogEntry on all the replicas (if the replica set is the same).
+// that way each replica knows who is responsible for doing a certain merge.
+
+// in some corner cases (added / removed / deactivated replica)
+// nodes can pick different replicas to execute merge and wait for it (or to execute the same merge together)
+// but that doesn't have a significant impact (in one case it will wait for the execute_merges_on_single_replica_time_threshold,
+// in another just 2 replicas will do the merge)
+std::optional<std::string> ReplicatedMergeTreeMergeStrategyPicker::pickReplicaToExecuteMerge(const ReplicatedMergeTreeLogEntryData & entry)
```

Comment:
we can return here smth like 
struct {
 strategy  enum (Merge, Fetch, Wait)
 picked_replica std::optional<std::string>
}

and incorpotate those fearutes here too: 
https://github.com/ClickHouse/ClickHouse/blob/026896835c7f48fc090dc3bec1dda6f023f61ef4/src/Storages/StorageReplicatedMergeTree.cpp#L1349-L1365 

https://github.com/ClickHouse/ClickHouse/blob/026896835c7f48fc090dc3bec1dda6f023f61ef4/src/Storages/StorageReplicatedMergeTree.cpp#L1392-L1410